osgar/drivers/eduro.py

```diff
@@ -0,0 +1,228 @@
+"""
+  Driver for robot Eduro
+"""
+
+
+import ctypes
+import serial
+import struct
+import math
+from threading import Thread
+
+from osgar.logger import LogWriter, LogReader
+from osgar.bus import BusShutdownException
+from .canserial import CAN_packet
+
+CAN_ID_SYNC = 0x80
+CAN_ID_ENCODERS_LEFT = 0x181
+CAN_ID_ENCODERS_RIGHT = 0x182
+CAM_SYSTEM_STATUS = 0x8A  # including emergency STOP button
+CAN_ID_BUTTONS = 0x28A
+
+UPDATE_TIME_FREQUENCY = 20.0  # Hz
+WHEEL_DIAMETER_LEFT = 427.0 / 445.0 * 0.26/4.0 + 0.00015
+WHEEL_DIAMETER_RIGHT = 427.0 / 445.0 * 0.26/4.0 - 0.00015
+
+
+def sint32_diff(a, b):
+    return ctypes.c_int32(a - b).value
+
+
+class Eduro(Thread):
+    def __init__(self, config, bus):
+        Thread.__init__(self)
+        self.setDaemon(True)
+
+        self.bus = bus
+
+        self.desired_speed = 0.0  # m/s
+        self.desired_angular_speed = 0.0
+        self.maxAcc = config.get('max_acc', 0.5)
+        self._rampLastLeft, self._rampLastRight = 0.0, 0.0
+        self.WHEEL_DISTANCE = 0.315        
+
+        self.prev_enc_raw = {}
+        self.dist_left_raw = 0
+        self.dist_right_raw = 0
+        self.dist_pose = 0, 0
+
+        self.emergency_stop = None  # uknown state
+        self.pose = [0.0, 0.0, 0.0]  # x, y in meters, heading in radians (not corrected to 2PI)
+        self.buttons = None  # unknown
+
+    def update_encoders(self, msg_id, data):
+        assert len(data) == 4, data
+        arr = struct.unpack('<i', data)
+        if msg_id in self.prev_enc_raw:
+            diff = sint32_diff(arr[0], self.prev_enc_raw[msg_id])
+
+            if msg_id == CAN_ID_ENCODERS_LEFT:
+                self.dist_left_raw += diff
+            elif msg_id == CAN_ID_ENCODERS_RIGHT:
+                self.dist_right_raw += diff
+        self.prev_enc_raw[msg_id] = arr[0]
+
+    def update_emergency_stop(self, msg_id, data):
+        assert len(data) == 8, len(data)
+        self.emergency_stop = (data[:2] == bytes([0,0x10])) and (data [3:] == bytes([0,0,0,0,0]))
+
+    def update_pose(self):
+        x, y, heading = self.pose
+
+        diff_left = self.dist_left_raw - self.dist_pose[0]
+        diff_right = self.dist_right_raw  - self.dist_pose[1]
+        self.dist_pose = self.dist_left_raw, self.dist_right_raw
```

Comment:
Why is `dist_pose` a tuple while `dist_left_raw` and `dist_right_raw` are separate variables? Is the inconsistency intentional?

Reply:
Could it be advantageous to store for each side just the current raw dist and the last diff? The previous value of the raw dist is only needed to calculate the diff. Storing the diff directly would be more straight forward.

Reply:
OK, it makes sense. Now encoders are reported as diff and that is also needed for `pose2d` computation. `self.dist_pose` was a bit of hack to get it running ... I will fix it.

Reply:
it looks much better, thanks ;)